Document when FE flow is applicable with regards to involved devices (#256)
camaraproject/IdentityAndConsentManagement: documentation/CAMARA-API-access-and-user-consent.md

- `API Exposure Platform`: the Operator's platform for exposing CAMARA APIs to ASPs and Aggregators, providing authentication and authorization mechanisms, and End-User Consent management. The API Exposure Platform typically consists of at least an Auth Server and an API Gateway.
-	`Application` or `Application Backend`: the ASP's software services that access CAMARA APIs.
- `Application Service Provider (ASP)`: the Legal Entity that provides the Application and/or services that consume CAMARA APIs.

> - `Authentication Device`: The device on which the user will authenticate/authorize the request, often a smartphone.
> Ref: https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#terminology

Reviewer: ```suggestion
- `Authentication Device`: The device on which the user will authenticate/authorize the request, often a smartphone.
```

-	`Resource Server`: the server that exposes protected resources to Applications. The Resource Server requires a valid access token to be provided before allowing access to the protected resource.
-	`Scope`: the OpenID Connect scope which maps one or more protected resources, some scopes may require processing of Personal Data.
- `Subscriber`: the mobile subscriber of the Operator. The Subscriber is usually also the End-User, but this is not always the case. For example, a parent may be the Subscriber of a mobile subscription for their child, the End-User.

> - `Target Device`: the device, which is the main resource (target) impacted by an API call and for which consent might need to be collected from resource owner.

Reviewer: To me the `Target Device` is the `Consumption Device` which is already defined in the document. And the access token will be associated the network authenticated identifier corresponding to the `Consumption Device`. It has been clarified as well [in the profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/46beef349cef154e46e68e500cc05b36f9184c04/documentation/CAMARA-Security-Interoperability.md?plain=1#L102) `...CAMARA REQUIRES API Providers to use network-based authentication in the Authorization Code Flow. Therefore, access tokens are issued for the network authenticated identifier.`

So, IMHO, you should refer to the `Consumption Device` instead. You may need somehow to make your point about the differences between Consumption and Target device clearer.

Reviewer: ```suggestion
- `Target Device`: the device that is the primary resource (target) affected by an API call, and for which Consent may need to be obtained from the Resource Owner.
```

Author: > To me the `Target Device` is the `Consumption Device` which is already defined in the document. And the access token will be associated the network authenticated identifier corresponding to the `Consumption Device`. It has been clarified as well [in the profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/46beef349cef154e46e68e500cc05b36f9184c04/documentation/CAMARA-Security-Interoperability.md?plain=1#L102) `...CAMARA REQUIRES API Providers to use network-based authentication in the Authorization Code Flow. Therefore, access tokens are issued for the network authenticated identifier.`
> 
> So, IMHO, you should refer to the `Consumption Device` instead.
> 
> You may need somehow to make your point about the differences between Consumption and Target device clearer.

@Jesus: I added the definition to the top of the document, applicable to FE (auth code) flow and CIBA. And it is exactly the point of the CR to state that in case of FE (auth code) flow we have consumption device == target device, but this is not necessarily the case in CIBA.

Note: In cases where Personal Data is processed by a CAMARA API, and Users can exercise their rights through mechanisms such as opt-in and/or opt-out, the use of Three-Legged Access Tokens is mandatory.

#### Authorization Code Flow (Frontend Flow)

> The authorization code flow is only applicable if the consumption device, which initiates the process of obtaining a Three-Legged-Access token from the Authorization Server, is also the target device of the intented Network API call(s) .

Reviewer: ```suggestion
The Authorization Code Flow is only applicable if the Consumption Device that initiates the process of obtaining a Three-Legged Access Token from the Authorization Server is also the Target Device of the intended CAMARA API call(s).
```

_NOTE: The technical ruleset is applicable only after a subproject has agreed to use a Three-Legged Access Token authentication flow. This ruleset provides a recommendation which will help API providers to align on the Three-Legged Access Token Flow and help with aggregation._

> If all API usecases point to the need of an 'On-Net' scenario and where the Consumption Device and Authentication Device are the same, the Frontend flow SHOULD be used. eg. NumberVerification. 

Reviewer: ```suggestion
If all API use cases point to the need of an 'On-Net' scenario and where the Consumption Device and Authentication Device are the same, the Frontend flow SHOULD be used e.g. Number Verification. 
```

Also note, the Authorization Code Flow is only applicable if the Consumption Device, that initiates the process of obtaining a Three-Legged Access Token from the Authorization Server, is also the Target Device of the intended CAMARA API call(s).

> This flow then enables On-Net scenarios where the mobile connection of the Consumption Device needs to be authenticated. for example, the [CAMARA Number Verification API](https://github.com/camaraproject/NumberVerification) due to the nature of its functionality where a User's phone number needs to be compared to the phone number associated with the mobile connection of the Consumption Device. 

Reviewer: ```suggestion
This flow then enables On-Net scenarios where the mobile connection of the Consumption Device needs to be authenticated. For example, the [CAMARA Number Verification API](https://github.com/camaraproject/NumberVerification) due to the nature of its functionality where a User's phone number needs to be compared to the phone number associated with the mobile connection of the Consumption Device. 
```

Reviewer: These are just typo corrections. I will commit the changes directly.

    - Check if User Consent is required as the lawful basis associated with the declared Scope and Purpose. 
      - If necessary, it will be checked in the operator's consent master whether user consent has already been given to the application for the user identifier and declared purpose.
    - If NOT granted, the operator performs the consent capture. Since the authorization code grant involves the interaction with application front-end, consent can be captured directly from the user through the application browser.
  - Covered scenarios: 
    - On-net (with mobile connection) & application front-end (with embedded browser)
    - Off-net scenarios using refresh_token, as long as there was a connection when the first access_token was requested.

>     - Note: Consumption device must be equal to target device of subsequent API call(s).

Reviewer: ```suggestion
    - Note: The Consumption Device must be equal to the Target Device of subsequent API call(s).
```
This would be already clarified in line 88, wouldn't be?

#### CIBA flow (Backend flow)

> The CIBA flow is applicable if the consumption device is equal to or different from the target device of the intented Network API call(s).

Reviewer: ```suggestion
The CIBA flow is applicable if the Consumption Device is equal to or different from the Target Device of the intended CAMARA API call(s).
```
I find this a bit confusing. May not be necessary...

> The CIBA flow is applicable if the Consumption Device is equal to or different from the Target Device of the intended CAMARA API call(s).

Reviewer: The applicability of the CIBA Flow is described in the `Technical ruleset for the Backend flow` section. It may be better to move this statement there and integrate it with the rest of the information in that section [here](https://github.com/camaraproject/IdentityAndConsentManagement/blob/bd65974fa496ec065c9da2acbda448e215c1bb9c/documentation/CAMARA-API-access-and-user-consent.md?plain=1#L272)

Something like...

```
If the API use case(s) indicate the need for an "Off-net" scenario and/or if the Consumption Device and Authentication Device are different, the Backend flow (CIBA) SHOULD be used.

Also note that in the case of the CIBA flow, it is applicable if the Consumption Device is the same or different from the Target Device of the intended CAMARA API call(s).
```

Author: agree, let's move

#### CIBA flow (Backend flow)

> The CIBA flow is applicable if the Consumption Device is equal to or different from the Target Device of the intended CAMARA API call(s).

Reviewer 2: CIBA is necessary if Consumption Device and Authentication Device are different. If Consumption Device and Authentication Device are the same, then OIDC Authentication Code Flow SHOULD be used.

The API Provider MUST send a message to the authentication device, identified by login_hint, because otherwise there would be no authentication at all. 

Note: In cases where personal data is processed by the API and users can exercise their rights through mechanisms such as opt-in and/or opt-out, the use of three-legged access tokens is mandatory. This ensures that the API remains in compliance with privacy regulations, upholding the principles of transparency and user-centric privacy-by-design.

User consent cannot be checked based on the login_hint value alone and without sending a message, because then CIBA would be a two-legged flow while according the the note above a three-legged flow is mandatory.

Author: **Relation between consumption device and authentication device:**
>CIBA is necessary if Consumption Device and Authentication Device are different. If Consumption Device and >Authentication Device are the same, then OIDC Authentication Code Flow SHOULD be used.

The API invoker cannot know whether consumption device and authentication device are different. The API invoker in general has no awareness at all about authentication devices. The API invoker only knows the target device and potentially also the consumption device. Who is the legal responsbile (authorization device) of the target device is only known to the backend - the telco operator.

The OIDC Authentication Code Flow protocol does not foresee to communicate whether the consumption device is the authentication device or not. It would be beneficial to get this information in the response, but this is not the case today.

**CIBA flow:**
>User consent cannot be checked based on the login_hint value alone and without sending a message, because then >CIBA would be a two-legged flow while according the the note above a three-legged flow is mandatory.
Can you please clarify this statement ? 
AFAIK CIBA transports the information about the target resource/target device in the login_hint, The consent check happens for the target resource specified in the login_hint, The API invoker does not have any knowlege about the consumption device in this context.

    - Check if user consent is required by lawful basis associated with the declared purpose. 
      - If necessary, it will be checked **in the operator's consent master** whether user consent has already been given to the application for the user identifier and declared purpose.
      - If NOT granted, **the operator’s consent capture procedure is triggered**. Out-of-band consent capture as part of asynchronous CIBA flow (e.g. push notification with fallback to SMS, etc...). **Operators can choose the consent capture mechanism that best suits their capabilities, preferences and needs**.
  - Covered scenarios: 
    - No front-end developer software in user device
    - Back-end services (e.g. bank BE anti-fraud validation using MSISDN).
    - Off-net scenarios (no mobile connection)
    - Device connected to WiFi
    - Device without UI (IoT)

>     - Note: Consumption device and target device of subsequent API call(s) can be different, access token is generated for the target device

Reviewer: I think this is already clarified by line 278 `If some use case(s) for an API point to "Off-net" scenarios and where Consumption Device and authentication devices could be different, the Backend flow should be used.`. Isn't it?

Reviewer: By definition, in an Auth code flow, the consumption device is the same as the authentication device. And CIBA is exactly the flow to use when this is not the case.

Author: > I think this is already clarified by line 278 `If some use case(s) for an API point to "Off-net" scenarios and where Consumption Device and authentication devices could be different, the Backend flow should be used.`. Isn't it?

No, this is NOT my point, I'm after the statement that the consumption device (where the app runs) and the TARGET device of an API call might be different in case of CIBA. Use case: "Find My Pet".
The the consumption device and the auth device might be different as well.
We have 3 different roles ("devices"): Consumption device (where the app runs), target device/resource (target of the API call), Authorization device: contact information identifying the legal responsible of a target device

Author: > By definition, in an Auth code flow, the consumption device is the same as the authentication device. And CIBA is exactly the flow to use when this is not the case.

I do NOT fully agree.
In Auth code flow, the consumption device is the same as the target device.
But whether the consumption device might give consent or not is a different question, e.g. Kid playing a game. Then the consumption device and the target device of the API call, e.g. for device location is the Kid's device, but the auth device might be the parent's phone.

So, I agree that we are addressing a limited set of use cases in CAMARA today. But we need to clarify the roles of the involved stakeholders, end users and resource owners. For simplicity named "consumption device", "target device", "authorization device".
